File: programme/ast-java/MainSolution.java
/* Project:  Vorlesung Uebersetzerbau
 * Descr:    Aufbau AST in JAVA
 * Kind:     Hauptprogramm
 * Author:   Prof. Dr. Juergen Vollmer <vollmer@dhbw-karlsruhe.de>
 * $Id: Main.java.in,v 1.8 2012/10/20 13:58:23 vollmer Exp $
 */
public class MainSolution
{
    public static void main(String argv[])
    {
	System.out.println("Construction of an AST for expressions");
	EXPR tree1 =
	     new BinaryExpr (
	 	 '+',
		 new IntConst (1),
		 new BinaryExpr (
		     '*',
		     new UnaryExpr (
                         '-',
		 	 new IntConst (2)),
		     new IntConst (3)));
	tree1.write();
	System.out.println("=" + tree1.eval());

	EXPR tree2 =
             new CondExpr (
	         new BinaryExpr (
	 	     '<',
		     new IntConst (2),
		     new IntConst (3)),
                 new IntConst (2),
                 new IntConst (3));
	tree2.write();
	System.out.println("=" + tree2.eval());

	System.out.println("List of expressions:");
	EXPRS list =
	      new Exprs (
		  new Exprs (
		      new NoExprs (),
		      tree2),
                  tree1);
	list.write_and_eval();

	System.out.println("Reverted list of expressions:");
	EXPRS r = list.reverse();
	r.write_and_eval();
    }
}

File: programme/ast-c/main-solution.c
/* Project:  Vorlesung Übersetzerbau
 * Descr:    Aufbau AST in C
 * Kind:     Hauptprogramm
 * Author:   Prof. Dr. Juergen Vollmer <vollmer@dhbw-karlsruhe.de>
 * $Id: main.c.in,v 1.3 2010/04/15 17:04:56 vollmer Exp $
 */

# include <stdio.h>

 # include "tree-solution.h"
 # include "write-solution.h"
 # include "eval-solution.h"

void write_and_eval (tTree tree)
{
  while (tree->Kind == kexprs) {
    write (tree->exprs.Expr);
    printf ("= %5.2f\n", eval(tree->exprs.Expr));
    tree = tree->exprs.Next;
  }
}

int main (int argc, char *argv[])
{
  tTree tree1, tree2, list;
  printf ("Construction of an AST for expressions\n");

  tree1 = mbinary_expr (
           '+',
	   mint_const (1),
	   mbinary_expr (
   	     '*',
	     mint_const (2),
	     mint_const (3)));
  write (tree1); printf ("= %5.2f\n", eval(tree1));

  tree2 = mcond_expr (
	  mbinary_expr (
            '<',
	    mint_const (2),
	    mint_const (3)),
	  mint_const (2),
	  mint_const (3));
  write (tree2); printf ("= %5.2f\n", eval(tree2));

  printf ("List of expressions:\n");
  list = mexprs (
          mexprs (
	   mno_exprs (),
           tree2),
         tree1);
  write_and_eval(list);

  printf ("Reverted list of expressions:\n");
  list = reverse (list);
  write_and_eval(list);

  return 0;
}
